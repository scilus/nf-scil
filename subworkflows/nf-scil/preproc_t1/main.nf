// ** Importing modules from nf-scil ** //
include { DENOISING_NLMEANS } from '../../../modules/nf-scil/denoising/nlmeans/main'
include { PREPROC_N4 } from '../../../modules/nf-scil/preproc/n4/main'
include { IMAGE_RESAMPLE } from '../../../modules/nf-scil/image/resample/main'
include { BETCROP_ANTSBET } from '../../../modules/nf-scil/betcrop/antsbet/main'
include { BETCROP_SYNTHBET} from '../../../modules/nf-scil/betcrop/synthbet/main'
include { BETCROP_CROPVOLUME as BETCROP_CROPVOLUME_T1 } from '../../../modules/nf-scil/betcrop/cropvolume/main'
include { BETCROP_CROPVOLUME as BETCROP_CROPVOLUME_MASK } from '../../../modules/nf-scil/betcrop/cropvolume/main'

workflow PREPROC_T1 {

    take:
        ch_image           // channel: [ val(meta), [ image ] ]
        ch_template        // channel: [ val(meta), [ template ] ]        , optional
        ch_probability_map // channel: [ val(meta), [ probability_map ] ] , optional
        ch_mask_nlmeans    // channel: [ val(meta), [ mask ] ]            , optional
        ch_ref_n4          // channel: [ val(meta), [ ref, ref_mask ] ]   , optional
        ch_ref_resample    // channel: [ val(meta), [ ref ] ]             , optional
        val_synth(false)   // value: (default: false)

    main:

        ch_versions = Channel.empty()

        // ** Denoising ** //
        ch_nlmeans = ch_image.join(ch_mask_nlmeans)
        DENOISING_NLMEANS ( ch_nlmeans )
        ch_versions = ch_versions.mix(DENOISING_NLMEANS.out.versions.first())

        // ** N4 correction ** //
        ch_N4 = DENOISING_NLMEANS.out.image.join(ch_ref_n4)
        PREPROC_N4 ( ch_N4 )
        ch_versions = ch_versions.mix(PREPROC_N4.out.versions.first())

        // ** Resampling ** //
        ch_resampling = PREPROC_N4.out.image.join(ch_ref_resample)
        IMAGE_RESAMPLE ( ch_resampling )
        ch_versions = ch_versions.mix(IMAGE_RESAMPLE.out.versions.first())

        // ** Brain extraction ** //
        if ( val_synth ) {
            ch_bet = IMAGE_RESAMPLE.out.image
            BETCROP_SYNTHBET ( ch_bet )
            ch_versions = ch_versions.mix(BETCROP_SYNTHBET.out.versions.first())
        }

        else {
            ch_template = ch_template.ifEmpty(Channel.error('Template is required for ANTS registration'))
            ch_probability_map = ch_probability_map.ifEmpty(Channel.error('Probability map is required for ANTS registration'))
            ch_bet = IMAGE_RESAMPLE.out.image.join(ch_template).join(ch_probability_map)
            BETCROP_ANTSBET ( ch_bet )
            ch_versions = ch_versions.mix(BETCROP_ANTSBET.out.versions.first())
        }

        // ** crop image ** //
        ch_crop = BETCROP_ANTSBET.out.t1.map{it + [[]]}
        BETCROP_CROPVOLUME_T1 ( ch_crop )
        ch_versions = ch_versions.mix(BETCROP_CROPVOLUME_T1.out.versions.first())

        // ** crop mask ** //
        ch_crop_mask = BETCROP_ANTSBET.out.mask.join(BETCROP_CROPVOLUME_T1.out.bounding_box)
        BETCROP_CROPVOLUME_MASK ( ch_crop_mask )
        ch_versions = ch_versions.mix(BETCROP_CROPVOLUME_MASK.out.versions.first())

    emit:
        image_nlmeans   = DENOISING_NLMEANS.out.image         // channel: [ val(meta), [ image ] ]
        image_N4        = PREPROC_N4.out.image                // channel: [ val(meta), [ image ] ]
        image_resample  = IMAGE_RESAMPLE.out.image            // channel: [ val(meta), [ image ] ]
        image_bet       = ch_bet_out.t1                       // channel: [ val(meta), [ t1 ] ]
        mask_bet        = ch_bet_out.out.mask                 // channel: [ val(meta), [ mask ] ]
        crop_box        = BETCROP_CROPVOLUME_T1.out.bounding_box // channel: [ val(meta), [ bounding_box ] ]
        mask_final      = BETCROP_CROPVOLUME_MASK.out.image   // channel: [ val(meta), [ mask ] ]
        t1_final        = BETCROP_CROPVOLUME_T1.out.image     // channel: [ val(meta), [ image ] ]
        versions        = ch_versions                         // channel: [ versions.yml ]
}
