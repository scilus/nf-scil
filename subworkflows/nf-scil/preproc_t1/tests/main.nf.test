nextflow_workflow {

    name "Test Subworkflow PREPROC_T1"
    script "../main.nf"
    workflow "PREPROC_T1"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/preproc_t1"

    tag "denoising/nlmeans"
    tag "preproc/n4"
    tag "image/resample"
    tag "betcrop/antsbet"
    tag "betcrop/synthbet"
    tag "betcrop/cropvolume"

    tag "load_test_data"

        setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "antsbet.zip" , "others.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("preproc_t1_classic") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        others: it.simpleName == "others"
                    }
                input[0] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_unaligned.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_template.nii.gz")
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]}
                input[3] = ch_split_test_data.others.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[4] = ch_split_test_data.others.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        []
                    ]}
                input[5] = ch_split_test_data.others.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("preproc_t1_option") {
        config "./nextflow_2.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        others: it.simpleName == "others"
                    }
                input[0] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_unaligned.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_template.nii.gz")
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]}
                input[3] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_unaligned_brainmask.nii.gz")
                    ]}
                input[4] = ch_split_test_data.others.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        []
                    ]}
                input[5] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_unaligned.nii.gz")
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

test("preproc_t1_synth") {
        config "./nextflow_synth.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        others: it.simpleName == "others"
                    }
                input[0] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_unaligned.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[3] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[4] = ch_split_test_data.others.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        []
                    ]}
                input[5] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[6] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
