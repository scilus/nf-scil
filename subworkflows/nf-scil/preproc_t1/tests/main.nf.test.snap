{
    "preproc_t1_option": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,ba910fdd1c9cc8f3d1c98b24dde522ef"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_n4.nii.gz:md5,f9b1e5305c8c176c8ef72168b8a33f2d"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__resampled.nii.gz:md5,a16a9d32732752b40f699cea2fb046bd"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet.nii.gz:md5,f0a2be8bc64bb8b0311bb7022e177623"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet_mask.nii.gz:md5,bdafc62e0e2ab35d0d48e1021d0a82db"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped_bbox.pkl:md5,39a3444e23d7ea6478ef1213f88e7715"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,02b2ca05a95c8494a85cda69ebe6d5e7"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped.nii.gz:md5,593d0b155da379bb41e8557ea5f9b92e"
                    ]
                ],
                "8": [
                    "versions.yml:md5,209daf2d60de195845dedbdf80c98afd",
                    "versions.yml:md5,2c0a6f978891003ebdf35c1ba060aa90",
                    "versions.yml:md5,6aecd5fc2de479eaa81fb30b248c647c",
                    "versions.yml:md5,97b6b9457caf964b9f20549672f692f0",
                    "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                    "versions.yml:md5,c6563545d67d7ab20ad849adcaafcb71"
                ],
                "crop_box": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped_bbox.pkl:md5,39a3444e23d7ea6478ef1213f88e7715"
                    ]
                ],
                "image_N4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_n4.nii.gz:md5,f9b1e5305c8c176c8ef72168b8a33f2d"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet.nii.gz:md5,f0a2be8bc64bb8b0311bb7022e177623"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,ba910fdd1c9cc8f3d1c98b24dde522ef"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__resampled.nii.gz:md5,a16a9d32732752b40f699cea2fb046bd"
                    ]
                ],
                "mask_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet_mask.nii.gz:md5,bdafc62e0e2ab35d0d48e1021d0a82db"
                    ]
                ],
                "mask_final": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,02b2ca05a95c8494a85cda69ebe6d5e7"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped.nii.gz:md5,593d0b155da379bb41e8557ea5f9b92e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,209daf2d60de195845dedbdf80c98afd",
                    "versions.yml:md5,2c0a6f978891003ebdf35c1ba060aa90",
                    "versions.yml:md5,6aecd5fc2de479eaa81fb30b248c647c",
                    "versions.yml:md5,97b6b9457caf964b9f20549672f692f0",
                    "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                    "versions.yml:md5,c6563545d67d7ab20ad849adcaafcb71"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-30T20:24:23.951944"
    },
    "preproc_t1_synth": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,b6d25b37b1101fea5c77bbbc711757c3"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_n4.nii.gz:md5,ef0d9af46aa1e330c5d7679ead500596"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__resampled.nii.gz:md5,88bae6bfeea1734111621b437e09548d"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bet_image.nii.gz:md5,80bf67681166f331e8495f359461b9d4"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__brain_mask.nii.gz:md5,df69d549d6919584a0f7b36113994296"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped_bbox.pkl:md5,fb2114f56d82f5f4465f6b204a8298eb"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,8fdb13e81f76139351c7b1056aebf1ed"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped.nii.gz:md5,6cd92a7c908816a47ae4146bfc6571c7"
                    ]
                ],
                "8": [
                    "versions.yml:md5,2c0a6f978891003ebdf35c1ba060aa90",
                    "versions.yml:md5,6aecd5fc2de479eaa81fb30b248c647c",
                    "versions.yml:md5,97b6b9457caf964b9f20549672f692f0",
                    "versions.yml:md5,9d515db259af59f33ccd14061fbc45d1",
                    "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                    "versions.yml:md5,c6563545d67d7ab20ad849adcaafcb71"
                ],
                "crop_box": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped_bbox.pkl:md5,fb2114f56d82f5f4465f6b204a8298eb"
                    ]
                ],
                "image_N4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_n4.nii.gz:md5,ef0d9af46aa1e330c5d7679ead500596"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bet_image.nii.gz:md5,80bf67681166f331e8495f359461b9d4"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,b6d25b37b1101fea5c77bbbc711757c3"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__resampled.nii.gz:md5,88bae6bfeea1734111621b437e09548d"
                    ]
                ],
                "mask_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__brain_mask.nii.gz:md5,df69d549d6919584a0f7b36113994296"
                    ]
                ],
                "mask_final": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,8fdb13e81f76139351c7b1056aebf1ed"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped.nii.gz:md5,6cd92a7c908816a47ae4146bfc6571c7"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2c0a6f978891003ebdf35c1ba060aa90",
                    "versions.yml:md5,6aecd5fc2de479eaa81fb30b248c647c",
                    "versions.yml:md5,97b6b9457caf964b9f20549672f692f0",
                    "versions.yml:md5,9d515db259af59f33ccd14061fbc45d1",
                    "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                    "versions.yml:md5,c6563545d67d7ab20ad849adcaafcb71"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-30T19:21:50.686836"
    },
    "preproc_t1_classic": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,b6d25b37b1101fea5c77bbbc711757c3"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_n4.nii.gz:md5,ef0d9af46aa1e330c5d7679ead500596"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__resampled.nii.gz:md5,88bae6bfeea1734111621b437e09548d"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet.nii.gz:md5,016f12c75e3543c27987a91eb2160a19"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet_mask.nii.gz:md5,cf0cc416ca8b352d3cba58d9132ab0da"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped_bbox.pkl:md5,5d9bd5d9c068031f34c34223b5892153"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,aee59cb0bf2a4e14bfe116dc68be365b"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped.nii.gz:md5,f4bea1286a90760eecf348dd9bffd2e6"
                    ]
                ],
                "8": [
                    "versions.yml:md5,209daf2d60de195845dedbdf80c98afd",
                    "versions.yml:md5,2c0a6f978891003ebdf35c1ba060aa90",
                    "versions.yml:md5,6aecd5fc2de479eaa81fb30b248c647c",
                    "versions.yml:md5,97b6b9457caf964b9f20549672f692f0",
                    "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                    "versions.yml:md5,c6563545d67d7ab20ad849adcaafcb71"
                ],
                "crop_box": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped_bbox.pkl:md5,5d9bd5d9c068031f34c34223b5892153"
                    ]
                ],
                "image_N4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_n4.nii.gz:md5,ef0d9af46aa1e330c5d7679ead500596"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet.nii.gz:md5,016f12c75e3543c27987a91eb2160a19"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,b6d25b37b1101fea5c77bbbc711757c3"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__resampled.nii.gz:md5,88bae6bfeea1734111621b437e09548d"
                    ]
                ],
                "mask_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet_mask.nii.gz:md5,cf0cc416ca8b352d3cba58d9132ab0da"
                    ]
                ],
                "mask_final": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,aee59cb0bf2a4e14bfe116dc68be365b"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_t1_cropped.nii.gz:md5,f4bea1286a90760eecf348dd9bffd2e6"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,209daf2d60de195845dedbdf80c98afd",
                    "versions.yml:md5,2c0a6f978891003ebdf35c1ba060aa90",
                    "versions.yml:md5,6aecd5fc2de479eaa81fb30b248c647c",
                    "versions.yml:md5,97b6b9457caf964b9f20549672f692f0",
                    "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                    "versions.yml:md5,c6563545d67d7ab20ad849adcaafcb71"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-30T20:05:33.018053"
    }
}